src/ComfortPay/Message/ChargeRequest.php

```diff
@@ -18,79 +26,172 @@ public function setTerminalId($value)
+            $this->validate('e2eReference');
+        }
+
+        if (!empty($data['submerchantId']) || !empty($data['location']) || !empty($data['city']) || !empty($data['alpha2CountryCode'])) {
+            $this->validate('submerchantId', 'location', 'city', 'alpha2CountryCode');
+        }
+
         return $data;
     }
 
     public function sendData($data)
     {
-        if ($this->getTestmode()) {
-            if (intval($data['cid']) % 2 == 0) {
-                return $this->response = new ChargeResponse($this, ['transactionStatus' => '02', 'transactionApproval' => '123']);
+        if ($this->getTestmode())
+        {
+            if ((int)$this->getReferedCardId() % 2 == 0) {
+                return $this->response = new ChargeResponse($this,
+                    ['transactionId' => $data['transactionId'], 'transactionStatus' => '02', 'transactionApproval' => '123']);
             }
-            return $this->response = new ChargeResponse($this, ['transactionStatus' => '00', 'transactionApproval' => '123']);
+            return $this->response = new ChargeResponse($this,
+                ['transactionId' => $data['transactionId'], 'transactionStatus' => '00', 'transactionApproval' => '123']);
+        }
+
+        $req = new \stdClass();
+        $req->transactionId = $data['transactionId'];
+        $req->referedCardId = $data['referedCardId'];
+        $req->merchantId = $data['merchantId'];
+        $req->terminalId = $data['terminalId'];
+        $req->amount = $data['amount'];
+        $req->parentTransactionId = $data['parentTransactionId'];
+        $req->cc = $data['cc'];
+
+        $transactionIdentificator = new \stdClass();
+
+        if (!empty($data['vs']) && !empty($data['ss'])) {
+            $symbols = new \stdClass();
+            $symbols->variableSymbol = $data['vs'];
+            $symbols->specificSymbol = $data['ss'];
+            $transactionIdentificator->symbols = $symbols;
+        } else {
+            $transactionIdentificator->e2eReference = $data['e2eReference'];
+        }
+
+        $req->transactionIdentificator = $transactionIdentificator;
+
+        if (!empty($data['submerchantId']) && !empty($data['location']) && !empty($data['city']) && !empty($data['alpha2CountryCode'])) {
+            $ipspData = new \stdClass();
+            $ipspData->submerchantId = $data['submerchantId'];
+            $ipspData->location = $data['location'];
+            $ipspData->city = $data['city'];
+            $ipspData->alpha2CountryCode = $data['alpha2CountryCode'];
+            $req->ipspData = $ipspData;
         }
 
+        $request = new \stdClass();
+        $request->req = $req;
+        $request->transactionType = $data['transactionType'];
+
         $client = $this->getSoapClient();
+        $response = $client->doCardTransaction($request);
 
-        $data = [
-            'transactionId' => $data['transactionId'],
-            'referedCardId' => $data['cid'],
-            'merchantId' => $data['ws'],
```

Comment:
The original implementation used `ws` as `merchantId`. I couldn't really find out why is that, but when I tried to use `ws` value provided by bank as regular `merchantId`, the signature for initial payment wouldn't match.

That means, that regular payment should use `merchantId` to identify merchant, but any backend charge (recurring payment) should use `ws` to identify merchant.